### PR TITLE
No float.. hmm not really :)

### PR DESCRIPTION
**Pull request by doniacld**
## 🚧 Changes

- Remove decimal part in Number and store only in integer and the power in the precision
- Refactor the tests removing decimal and updating the integer parts

## ❓ Questions
- Should we round or truncate? (See TODO in the code)

### CODE REVIEW

File: chapter-05/final/money/convert_test.go [hunk 1]
```diff
@@ -69,6 +68,8 @@ func TestConvert(t *testing.T) {
 				}
 			},
 		},
+		// TODO FIX ME, not too small cause we are rouding on the float value.
```
Comment (doniacld) [inline chapter-05/final/money/convert_test.go:71]: TODO FIX ME

File: chapter-05/final/money/convert_test.go [hunk 2]
```diff
@@ -81,23 +82,22 @@ func TestConvert(t *testing.T) {
...
-					t.Errorf("expected error %s, got %v", money.ErrUnknownChangeRate, err)
-				}
-			},
-		},
 	}
 
 	for name, tc := range tt {
 		t.Run(name, func(t *testing.T) {
-			got, err := money.Convert(context.Background(), tc.amount, tc.from, tc.to, tc.rateRepo)
+			// TODO not sure we should do the parsing here but fields from Number and Currency are not exposed
```
Comment (doniacld) [inline chapter-05/final/money/convert_test.go:89]: TODO: I am not sure if we should do the parsing here in the tests but it was easier to write the tests case. Maybe I was too lazy? 
We have a NewCurrency as helper should we do the same for Number and just remove the parsing?

File: chapter-05/final/money/decimal.go [hunk 3]
```diff
@@ -12,61 +13,81 @@ import (
...
+
+	return Number{integerPart: number, precision: precision}, nil
 }
 
 // applyChangeRate returns a new Number representing n multiplied by the rate.
 // The precision is the same in and out.
 func (n Number) applyChangeRate(rate ChangeRate, toPrecision int) Number {
+	// compute the new amount
 	converted := n.float() * float64(rate)
+	// shift the period to handle an integer
```
Comment (floppyzedolfin) [inline chapter-05/final/money/decimal.go:71]:
```suggestion
	// shift the decimal separator to handle an integer
```

File: chapter-05/final/money/decimal.go [hunk 4]
```diff
@@ -12,61 +13,81 @@ import (
...
+	inFloat := converted * math.Pow10(toPrecision)
 
-	floor := math.Floor(converted)
-	decimal := math.Round((converted - floor) * math.Pow10(toPrecision))
+	// TODO: I am not sure if we should round or trunc here?
```
Comment (floppyzedolfin) [inline chapter-05/final/money/decimal.go:74]: trunc.
It's dangerous to round  and also not intuitive.

If you are in charge of a bureau of change, and someone gives you X that converts to 42.576, you are not going to give 42.58, but instead you'll return 42.57. In the first case, you gave/lost 0.004 from your own money, while in the second case, the customer paid/lost the 0.006.

Think of it this way:
If I convert X to the new currency, and the output is 1.0057, I can either floor it to 1.00 or round it to 1.01.
If 10000 people do that change, I will either have given away 10000, or I will have given away 10100. In the first case, my company has "won" 57 (by round to 1.00), and in the second case, it has lost 43 (by rounding to 1.00).

File: chapter-05/final/money/decimal.go [hunk 5]
```diff
@@ -78,9 +99,9 @@ const (
 	ErrInputTooLarge = moneyError("input amount over 10^15 is too large")
 	// ErrOutputTooLarge is returned if the converted amount is too large, to protect against floating point errors.
 	ErrOutputTooLarge = moneyError("output amount is too large (over 10^15)")
+	// ErrInputTooPrecise is returned if the input amount is too precise, which could lead to precision issues.
+	ErrInputTooPrecise = moneyError("input amount should be at least 1 * precision")
```
Comment (floppyzedolfin) [inline chapter-05/final/money/decimal.go:103]: precision is something internal to the system. It shouldn't be exposed. "input includes subdivisions of the currency that are too small"

**Comment by doniacld** [comment]
This PR is obsolete because we are now dealing with float64.

**Pull request closed without merging**